### PR TITLE
feat: compute WriteSummary from PersistenceWindows (#2030)

### PR DESCRIPTION
**Pull request by tustvold**
Adds a new `WriteSummary` data_type and adds the ability to generate it from PersistenceWindows. This is a precursor to wiring up a system table as part of #2030

### CODE REVIEW

File: persistence_windows/src/persistence_windows.rs [hunk 1]
```diff
@@ -335,6 +350,30 @@ impl PersistenceWindows {
         self.windows().next()
     }
 
+    /// Returns approximate summaries of the writes contained within this PersistenceWindow
```
Comment (alamb) [inline persistence_windows/src/persistence_windows.rs:353]: Can we try and explain what is "approximate" about the summary? I would be confused about what was wrong (aka if it is approximate what could it be used for?) if I came across this comment in the codebase

File: persistence_windows/src/persistence_windows.rs [hunk 2]
```diff
@@ -1265,4 +1314,89 @@ mod tests {
...
+
+        w.add_range(
+            Some(&Sequence { id: 1, number: 3 }),
+            6,
+            Utc.timestamp_nanos(1),
+            Utc.timestamp_nanos(5),
+            instant + Duration::from_millis(50),
+        );
+
+        // Window 2
```
Comment (alamb) [inline persistence_windows/src/persistence_windows.rs:1348]: what makes this a second window? The fact that it is  `DEFAULT_CLOSED_WINDOW_PERIOD` later than `instant`?

**Review by alamb — APPROVED** [review]
LGTM -- I had some minor suggestions regarding docs, etc but the code looks good 👍

File: persistence_windows/src/persistence_windows.rs [hunk 3]
```diff
@@ -106,14 +116,19 @@ impl PersistenceWindows {
 
         let closed_window_count = late_arrival_seconds / closed_window_seconds;
 
+        let created_at_time = Utc::now();
+        let created_at_instant = Instant::now();
```
[the remark this answers is not in the document]
Comment (tustvold) [reply persistence_windows/src/persistence_windows.rs:120]: Instants are guaranteed to be monotonic, UTC timestamps are not. The PersistenceWindows require monotonically increasing time, and so we must use Instant. However, if we want to display we must convert this to a UTC timestamp. In general using UTC wall clock timestamps for decision making is likely a bad idea.

File: persistence_windows/src/persistence_windows.rs (hunk 2 above)
Comment (tustvold) [reply persistence_windows/src/persistence_windows.rs]: I added comments clarifying

File: persistence_windows/src/persistence_windows.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (alamb) [reply persistence_windows/src/persistence_windows.rs:120]: So following that logic, perhaps we should use only instants internally and convert to UTC timestamps when displaying?

[the remark this answers is not in the document]
Comment (tustvold) [reply persistence_windows/src/persistence_windows.rs:120]: I think that would be reasonable for any timestamps that are used for anything other than display, and don't need to be persisted. As it stands I think PersistenceWindows is the only place that contains wall-clock based logic, but I could be wrong on that.

[the remark this answers is not in the document]
Comment (tustvold) [reply persistence_windows/src/persistence_windows.rs:120]: Perhaps... Not sure how complex a change that would be though, also not entirely sure how persistence would work.

**Pull request merged**